library/alloc/src/rc.rs

```diff
@@ -933,6 +934,29 @@ impl<T: ?Sized> Rc<T> {
         Weak { ptr: this.ptr }
     }
 
+    /// Convert a reference to an [`Rc`] into a reference to a [`Weak`] of the same type.
+    ///
+    /// This is a type-only operation; it doesn't modify the inner reference counts.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(rc_as_weak)]
+    ///
+    /// use std::rc::{Rc, Weak};
+    ///
+    /// let five: &Rc<i32> = &Rc::new(5);
+    ///
+    /// let weak_five: &Weak<i32> = Rc::as_weak(five);
+    /// ```
+    #[inline]
+    #[unstable(feature = "rc_as_weak", issue = "100472")]
+    #[must_use]
+    pub const fn as_weak(this: &Self) -> &Weak<T> {
+        let weak = this as *const Self as *const Weak<T>;
+        unsafe { &*weak }
```

Comment:
Can you add a SAFETY comment explaining why this is ok? https://std-dev-guide.rust-lang.org/code-considerations/safety-and-soundness/summary.html

Reply:
Done. I'm open to suggestions for improved wording. 🙃